src/inferencer/types.ts

```diff
@@ -0,0 +1,300 @@
+  }
+
+  /**
+   *
+   * @param description the description of this constrained type variable
+   * @param possibleTypes the types that this type variable can contain
+   */
+  constructor(
+    private description: string,
+    public readonly possibleTypes: (new (...args: any[]) => Type)[]
+  ) {
+    super()
+  }
+
+  /**
+   * Checks to see if candidate is of the correct type.
+   * @param candidate
+   */
+  canContain(candidate: Type): boolean {
+    return this.possibleTypes.some(possibleType => candidate instanceof possibleType)
+  }
+}
+
+/**
+ * The Integer type, uses `Number.isInteger` to check if integer.
+ *
+ * Implemented using a self-referential constrained type of either Float or Integer.
+ *
+ * If uninitialised, we know that it's an integer. *
+ * If initialised, it was initialised to either another integer or a float, so we call the toString of the instance.
+ */
+export class Number extends Constrained {
+  constructor() {
+    super('number', [Float, Integer, Number])
+  }
+
+  toString(): string {
+    if (this.instance === null) {
+      return this.name
+    } else {
+      return this.instance.toString()
+    }
+  }
+}
+
+/**
+ * The "Comparable" type, used for primitive operations that work on numbers/strings
+ * such as + < > <= >=.
+ */
+export class Comparable extends Constrained {
+  constructor() {
+    super('comparable', [String, Float, Integer, Number, Comparable])
+  }
+}
+
+/**
+ * The polymorphic type
+ */
+export class Polymorphic implements Type {
+  constructor(public name: string, public types: Type[]) {}
+
+  toString() {
+    return `${this.name}<${this.types.join(', ')}>`
+  }
+
+  fresh(mappings: Map<string, Type> = new Map()): Type {
+    const cloned = clone(this)
+    cloned.types = cloned.types.map((elementType: Type) => elementType.fresh(mappings))
+    return cloned
+  }
+}
+
+/**
+ * The type of a general n-ary function, which captures a n-tuple of types of its parameter, and its return type.
+ */
+export class Function extends Polymorphic {
+  /**
+   * @param _argTypes The types of the arguments of this function.
+   * @param _retType  The return type of this function.
+   */
+  constructor(private _argTypes: Type[], private _retType: Type) {
+    super('Function', [..._argTypes, _retType])
+  }
+  toString() {
+    const argStr = this._argTypes.map(t => t.toString()).join(', ')
+    const separator = ' -> '
+    return this._argTypes.length === 1
+      ? `(${argStr}${separator}${this._retType.toString()})`
+      : `((${argStr})${separator}${this._retType.toString()})`
+  }
+
+  get argTypes() {
+    return this.types.slice(0, -1)
+  }
+
+  get retType() {
+    return this.types[this.types.length - 1]
+  }
+}
+
+/**
+ * The polymorphic List type
+ */
+export class List extends Polymorphic {
+  constructor(v: Type) {
+    super('List', [v])
+  }
+}
+
+export class Pair extends Polymorphic {
+  constructor(head: Type, tail: Type) {
+    super('Pair', [head, tail])
+  }
+}
+
```

Comment:
Wanted to ask why should the function, list and pair types be polymorphic by default? I believe that they should be monomorphic by default and then polymorphism is like some sort of wrapper that wraps around any monomorphic type.

Reply:
I guess it should be named Parameterised instead to reduce confusion?